cli/cmds/cluster_update.go

```diff
@@ -0,0 +1,166 @@
+		}
+
+		namespace := appCtx.Namespace(name)
+
+		var virtualCluster v1beta1.Cluster
+
+		clusterKey := types.NamespacedName{Name: name, Namespace: appCtx.namespace}
+		if err := appCtx.Client.Get(ctx, clusterKey, &virtualCluster); err != nil {
+			return fmt.Errorf("failed to fetch existing cluster: %w", err)
+		}
+
+		if config.version != "" {
+			currentVersion := virtualCluster.Spec.Version
+			if currentVersion == "" {
+				currentVersion = virtualCluster.Status.HostVersion
+			}
+
+			currentVersionSemver, err := semver.ParseTolerant(currentVersion)
+			if err != nil {
+				return fmt.Errorf("failed to parse current cluster version %w", err)
+			}
+
+			newVersionSemver, err := semver.ParseTolerant(config.version)
+			if err != nil {
+				return fmt.Errorf("failed to parse new cluster version %w", err)
+			}
+
+			if newVersionSemver.LT(currentVersionSemver) {
+				return fmt.Errorf("downgrading cluster version is not supported")
+			}
+		}
+
+		logrus.Infof("Updating cluster '%s' in namespace '%s'", name, namespace)
+
+		virtualCluster.Spec.Servers = ptr.To(int32(config.servers))
+		virtualCluster.Spec.Agents = ptr.To(int32(config.agents))
+		virtualCluster.Spec.ServerArgs = config.serverArgs
+		virtualCluster.Spec.AgentArgs = config.agentArgs
+		virtualCluster.Spec.Version = config.version
+		virtualCluster.Labels = parseKeyValuePairs(config.labels, "label")
+		virtualCluster.Annotations = parseKeyValuePairs(config.annotations, "annotation")
+
+		if err := client.Update(ctx, &virtualCluster); err != nil {
+			return err
+		}
+
+		clusterDetails, err := printClusterDetails(&virtualCluster)
+		if err != nil {
+			return fmt.Errorf("failed to print cluster details: %w", err)
+		}
+
+		logrus.Info(clusterDetails)
+
+		logrus.Infof("Waiting for cluster to be available..")
+
+		if err := waitForClusterReady(ctx, client, &virtualCluster, config.timeout); err != nil {
+			return fmt.Errorf("failed to wait for cluster to become ready (status: %s): %w", virtualCluster.Status.Phase, err)
+		}
+
+		logrus.Infof("Extracting Kubeconfig for '%s' cluster", name)
+
+		// retry every 5s for at most 2m, or 25 times
+		availableBackoff := wait.Backoff{
+			Duration: 5 * time.Second,
+			Cap:      2 * time.Minute,
+			Steps:    25,
+		}
+
+		// add Host IP address as an extra TLS-SAN to expose the k3k cluster
+		url, err := url.Parse(appCtx.RestConfig.Host)
+		if err != nil {
+			return err
+		}
+
+		host := strings.Split(url.Host, ":")
+		if config.kubeconfigServerHost != "" {
+			host = []string{config.kubeconfigServerHost}
+		}
+
+		virtualCluster.Spec.TLSSANs = []string{host[0]}
+
+		cfg := kubeconfig.New()
+
+		var kubeconfig *clientcmdapi.Config
+
+		if err := retry.OnError(availableBackoff, apierrors.IsNotFound, func() error {
+			kubeconfig, err = cfg.Generate(ctx, client, &virtualCluster, host[0], 0)
+			return err
+		}); err != nil {
+			return err
+		}
+
+		return writeKubeconfigFile(&virtualCluster, kubeconfig, "")
```

Comment:
As already discussed I don't think we should wait for the cluster to be ready. Some of the fields are not going to actually change something that needs a restart (i.e. labels/annotations), and I think it's better to have a fire&forget approach, and eventually add a `--w/wait` flag to wait for the update.

(We can think about an `--immediate` flag on the `cluster create` eventually).